Source/Core/VideoBackends/D3D/PostProcessing.cpp

```diff
@@ -0,0 +1,724 @@
+	g_renderer->ResetAPIState();
+
+	// Copied from Renderer::SetViewport
+	int scissorXOff = bpmem.scissorOffset.x * 2;
+	int scissorYOff = bpmem.scissorOffset.y * 2;
+
+	float X = Renderer::EFBToScaledXf(xfmem.viewport.xOrig - xfmem.viewport.wd - scissorXOff);
+	float Y = Renderer::EFBToScaledYf(xfmem.viewport.yOrig + xfmem.viewport.ht - scissorYOff);
+	float Wd = Renderer::EFBToScaledXf(2.0f * xfmem.viewport.wd);
+	float Ht = Renderer::EFBToScaledYf(-2.0f * xfmem.viewport.ht);
+	if (Wd < 0.0f)
+	{
+		X += Wd;
+		Wd = -Wd;
+	}
+	if (Ht < 0.0f)
+	{
+		Y += Ht;
+		Ht = -Ht;
+	}
+
+	// In D3D, the viewport rectangle must fit within the render target.
+	TargetRectangle target_rect;
+	TargetSize target_size(g_renderer->GetTargetWidth(), g_renderer->GetTargetHeight());
+	target_rect.left = static_cast<int>((X >= 0.f) ? X : 0.f);
+	target_rect.top = static_cast<int>((Y >= 0.f) ? Y : 0.f);
+	target_rect.right = target_rect.left + static_cast<int>((X + Wd <= g_renderer->GetTargetWidth()) ? Wd : (g_renderer->GetTargetWidth() - X));
+	target_rect.bottom = target_rect.bottom + static_cast<int>((Y + Ht <= g_renderer->GetTargetHeight()) ? Ht : (g_renderer->GetTargetHeight() - Y));
+
+	// Source and target textures, if MSAA is enabled, this needs to be resolved
+	D3DTexture2D* color_texture = FramebufferManager::GetResolvedEFBColorTexture();
+	D3DTexture2D* depth_texture = nullptr;
+	if (m_requires_depth_buffer)
+		depth_texture = FramebufferManager::GetResolvedEFBDepthTexture();
+
+	// Invoke post process process
+	PostProcess(target_rect, target_size, FramebufferManager::GetEFBLayers(), reinterpret_cast<uintptr_t>(color_texture), reinterpret_cast<uintptr_t>(depth_texture));
+
+	// Copy back to EFB buffer when multisampling is enabled
+	if (g_ActiveConfig.iMultisamples > 1)
+		CopyTexture(target_rect, FramebufferManager::GetEFBColorTexture(), target_rect, color_texture, target_size, -1, true);
+
+	g_renderer->RestoreAPIState();
+
+	// Restore EFB target
+	D3D::context->OMSetRenderTargets(1, &FramebufferManager::GetEFBColorTexture()->GetRTV(),
+		FramebufferManager::GetEFBDepthTexture()->GetDSV());
+}
+
+void D3DPostProcessor::BlitToFramebuffer(const TargetRectangle& dst_rect, const TargetSize& dst_size, uintptr_t dst_texture,
+										 const TargetRectangle& src_rect, const TargetSize& src_size, uintptr_t src_texture,
+										 int src_layer)
+{
+	D3DTexture2D* real_dst_texture = reinterpret_cast<D3DTexture2D*>(dst_texture);
+	D3DTexture2D* real_src_texture = reinterpret_cast<D3DTexture2D*>(src_texture);
+	_dbg_assert_msg_(VIDEO, src_layer >= 0, "BlitToFramebuffer should always be called with a single source layer");
+
+	// Check for changed options.
+	if (m_scaling_shader)
+	{
+		if (!m_scaling_shader->IsReady() || !m_scaling_shader->Reconfigure(src_size))
+			m_scaling_shader.reset();
+		else
+			m_scaling_config->ClearDirty();
+	}
+
+	// Use scaling shader if one is set-up. Should only be a single pass in almost all cases.
```

Comment:
More advanced scaling shaders like xBR, Super xBR and NEDI are multi-pass.
https://github.com/libretro/common-shaders/tree/master/xbr/shaders
https://github.com/libretro/common-shaders/tree/master/nedi/shaders